### PR TITLE
Associated consts are not object safe.

### LINKED ISSUE #26847: Associated constants should not be object safe
The following code compiles:

``` rust
#![feature(associated_consts)]

trait Foo {
    const FOO: u32;
}

impl Foo for () {
    const FOO: u32 = 1;
}

fn main() {
    let _: &Foo = &();
}
```

Furthermore, adding the line `<Foo>::FOO;` to `main` causes an ICE:

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: match trait_items[method_offset_in_trait] {
    ty::MethodTraitItem(_) => true,
    _ => false,
}', ../src/librustc/middle/traits/util.rs:439
```

### PR DESCRIPTION
**Pull request by withoutboats**
fixes #26847

r? @eddyb

### CODE REVIEW

**Comment by withoutboats** [comment]
added compile-fail test

File: src/test/compile-fail/object-safety-associated-consts.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// from traits with associated consts.
+
+trait Bar {
+    const X: usize;
+}
+
+fn make_bar<T:Bar>(t: &T) -> &Bar {
+        //~^ ERROR E0038
+        //~| NOTE the trait cannot contain associated consts like `X`
+        //~| NOTE the trait `Bar` cannot be made into an object
```
Comment (eddyb) [inline src/test/compile-fail/object-safety-associated-consts.rs]: Unnecessarily indented once to the right.

**Pull request merged**